src/common/options.cc

```diff
@@ -5981,6 +5981,14 @@ std::vector<Option> get_mds_client_options() {
     .set_default(true)
     .set_description(""),
 
+    Option("client_deleg_timeout", Option::TYPE_FLOAT, Option::LEVEL_ADVANCED)
+    .set_default(75.0)
+    .set_description("time (in seconds) that application has to return delegation after recall"),
+
+    Option("client_deleg_break_on_open", Option::TYPE_BOOL, Option::LEVEL_ADVANCED)
+    .set_default(false)
```

Comment:
My thinking was that the application (ganesha and probably samba) could just call:

ceph_conf_set("client_deleg_break_on_open", "true");

whenever it creates a client. Since this is all about what sort of behaviour the application wants, I think that the simplest approach that doesn't change, e.g. ceph-fuse.

Reply:
So what happens if we have a native CephFS client that doesn't know about these? Is it okay for that to open a file without delegations being broken, as long as it doesn't do any reads or writes?
I thought we needed to be able to actually deny opens until delegations are broken. And that means every client in the cluster, regardless of if they know about NFS directly, needs to have the same behavior.

Reply:
It really comes down to applications that need to enforce these sort of semantics. We can't  enforce deny mode locking cluster-wide since it's not supported there in any fashion.

FWIW, POSIX doesn't really have any sort of share/deny mode locking (though apparently BSD has O_SHLOCK and O_EXLOCK, which is close). NFS and SMB both do however, and need to enforce those semantics. So, we need this behavior in them at least, or we can't handle the case of a cached deny mode lock on the client. For other callers, it not as crucial:

Suppose we have a server application (like ganesha or samba) that has a deny lock set such that another client would be denied an open, and then a "normal" ceph client (e.g. ceph-fuse or whatever) comes along and opens the file?

He'll get an open and eventually be able to do I/O to the file, like normal. This is primarily to ensure that we can at least enforce it within a single application (ganesha or samba). Eventually we may need to add something for cluster-coherent deny locks, but I'm not proposing any mechanism like that here.